examples/model_serving.py

```diff
@@ -0,0 +1,145 @@
+   e.g. initializing NNs. An easy way to ensure this is to inherit from
+   `nn.Module`, and do this initialization in `__init__`.
+ - Likewise, have a separate prediction function that derives from `nn.Module`.
+   This allows us to store parameters as `nn.Parameter` attributes that can be
+   serialized by `torch.jit.save`. For SVI, we can also simply store the learned
+   parameter values as constants in the trace. Note that we use `torch.jit.trace_module`
+   (instead of scripting), which supports many of the constructs needed for running
+   doing predictions with Pyro models, like context managers and higher order functions.
+
+**References:**
+
+  - TorchScript FAQs, https://pytorch.org/docs/stable/jit.html#frequently-asked-questions
+ """
+
+import argparse
+import logging
+import os
+
+import numpy as np
+import pandas as pd
+import torch
+
+from pyro.infer import SVI, Trace_ELBO, JitTrace_ELBO
+from pyro.optim import Adam
+from pyro import poutine
+from torch import nn
+
+import pyro
+import pyro.distributions as dist
+from pyro.infer.autoguide import AutoDiagonalNormal
+from pyro.poutine.runtime import param_no_grad
+
+DATA_URL = "https://d2hg8soec8ck9v.cloudfront.net/datasets/rugged_data.csv"
+DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '.data')
+logging.basicConfig(format='%(relativeCreated) 9d %(message)s', level=logging.INFO)
+
+
+class Model(nn.Module):
+    def __init__(self, p):
+        super(Model, self).__init__()
+        self.linear = nn.Linear(p, 1)
+        weight_shape, bias_shape = self.linear.weight.shape, self.linear.bias.shape
+        self.priors = {
+            "weight": dist.Normal(torch.zeros(weight_shape), torch.ones(weight_shape)).to_event(2),
+            "bias": dist.Normal(8. * torch.ones(bias_shape), 1000.)
+        }
+
+    def forward(self, x, y=None):
+        lifted_module = pyro.random_module("reg_module", self.linear, self.priors)()
+        sigma = pyro.sample("sigma", dist.Uniform(0., 10.))
+        with pyro.plate("data", x.shape[0]):
+            mean = lifted_module(x).squeeze(-1)
+            return pyro.sample("obs", dist.Normal(mean, sigma), obs=y)
+
+
+class Guide(nn.Module):
+    def __init__(self, model, p):
+        # Limit the scale parameter to stabilize learning.
+        pyro.param("auto_scale", torch.ones(p + 2) * 1e-4,
+                   constraint=dist.constraints.positive)
+        self.autoguide = AutoDiagonalNormal(model)
+        super(Guide, self).__init__()
+
+    def forward(self, x, y=None):
+        return self.autoguide(x, y)
+
+
+class Predictor(nn.Module):
+    def __init__(self, model, guide):
+        super(Predictor, self).__init__()
+        self.model = model
+        self.guide = guide
+        self.params = {}
+
+    def forward(self, x, y):
+        # Replay using saved guide parameters
+        guide_trace = poutine.trace(poutine.replay(self.guide, params=self.params)).get_trace(x, y=y)
```

Comment:
Sorry, that's left over from my previous commit where I was storing `self.params` as `nn.Parameters`. Its not needed now since we are simply storing these parameters as constants in the trace. 

> Will it be simpler if we get params from param stores, store non-grad version in Predictor, then jit

I was doing exactly that in my previous commit, but its not needed for SVI. It might still be better to do it explicitly so that we can reuse the same pattern for NUTS. I'll try to keep that in mind as I get the NUTS predictive version working. 

> but I can't figure out why param_no_grad is needed. 

I think param initialization results in creating tensors that have `requires_grad=True` and JIT tracing complains about not having constants that require grad, even if these constants are not eventually used.

Reply:
Understood! Thanks, Neeraj!